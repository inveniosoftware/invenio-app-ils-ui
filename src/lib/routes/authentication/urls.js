import { generatePath } from 'react-router-dom';
const AuthenticationBase = '/';

const AuthenticationRoutesList = {
  login: `${AuthenticationBase}login`,
  redirectUrlAfterLogin: `${AuthenticationBase}login?next=:nextUrl`,
  confirmEmail: `${AuthenticationBase}confirm-email`,
};

const AuthenticationRoutesGenerators = {
  redirectAfterLogin: (nextUrl) =>
    generatePath(AuthenticationRoutesList.redirectUrlAfterLogin, {
      nextUrl: nextUrl,
    }),
};

export const AuthenticationRoutes = {
  ...AuthenticationRoutesList,
  ...AuthenticationRoutesGenerators,
};
