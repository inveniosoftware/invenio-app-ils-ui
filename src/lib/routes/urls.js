export { FrontSiteRoutes } from '@routes/frontsite/frontsiteUrls';
export { BackOfficeRoutes } from '@routes/backoffice/backofficeUrls';
export { AuthenticationRoutes } from './authentication/urls';
export { AcquisitionRoutes } from './backoffice/AcquisitionUrls';
export { ProviderRoutes } from './backoffice/ProvidersUrls';
export { ILLRoutes, DetailsRouteByPidTypeFor } from './backoffice/ILLUrls';
