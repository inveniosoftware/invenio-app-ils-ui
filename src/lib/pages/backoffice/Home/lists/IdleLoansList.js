import { dateFormatter, toShortDate } from '@api/date';
import { loanApi } from '@api/loans';
import { withCancel } from '@api/utils';
import { SeeAllButton } from '@components/backoffice/buttons/SeeAllButton';
import { Error } from '@components/Error';
import { Loader } from '@components/Loader';
import { ResultsTable } from '@components/ResultsTable/ResultsTable';
import { invenioConfig } from '@config';
import LiteratureTitle from '@modules/Literature/LiteratureTitle';
import { BackOfficeRoutes } from '@routes/urls';
import { DateTime } from 'luxon';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Button } from 'semantic-ui-react';

export class IdleLoansList extends Component {
  constructor(props) {
    super(props);

    this.query = loanApi
      .query()
      .withState(invenioConfig.CIRCULATION.loanRequestStates)
      .withUpdated({
        to: toShortDate(DateTime.local().minus({ days: 10 })),
      })
      .qs();

    this.state = {
      data: { hits: [], total: 0 },
      isLoading: true,
      error: {},
    };
  }

  componentDidMount() {
    this.fetch();
  }

  componentWillUnmount() {
    this.cancellableFetch && this.cancellableFetch.cancel();
  }

  fetch = async () => {
    this.cancellableFetch = withCancel(loanApi.list(this.query));
    this.setState({ isLoading: true });
    try {
      const response = await this.cancellableFetch.promise;
      this.setState({ data: response.data, isLoading: false, error: {} });
    } catch (error) {
      if (error !== 'UNMOUNTED') {
        this.setState({ isLoading: false, error: error });
      }
    }
  };

  seeAllButton = () => {
    const path = BackOfficeRoutes.loansListWithQuery(this.query);
    return <SeeAllButton to={path} />;
  };

  viewDetails = ({ row }) => {
    return (
      <Button
        as={Link}
        to={BackOfficeRoutes.loanDetailsFor(row.metadata.pid)}
        compact
        icon="info"
        data-test={row.metadata.pid}
      />
    );
  };

  renderTable(data) {
    const { showMaxEntries } = this.props;
    const columns = [
      { title: '', field: '', formatter: this.viewDetails },
      { title: 'ID', field: 'metadata.pid' },
      { title: 'Patron', field: 'metadata.patron.name' },
      {
        title: 'Title',
        field: '',
        formatter: ({ row }) => (
          <LiteratureTitle
            title={row.metadata.document.title}
            edition={row.metadata.document.edition}
            publicationYear={row.metadata.document.publication_year}
          />
        ),
      },
      {
        title: 'Request start date',
        field: 'metadata.request_start_date',
        formatter: dateFormatter,
      },
    ];
    return (
      <ResultsTable
        data={data.hits}
        columns={columns}
        totalHitsCount={data.total}
        title="Idle loan requests"
        subtitle="Loan requests pending since more than 10 days."
        name="idle loan requests"
        seeAllComponent={this.seeAllButton()}
        showMaxRows={showMaxEntries}
      />
    );
  }

  render() {
    const { data, isLoading, error } = this.state;
    return (
      <Loader isLoading={isLoading}>
        <Error error={error}>{this.renderTable(data)}</Error>
      </Loader>
    );
  }
}

IdleLoansList.propTypes = {
  showMaxEntries: PropTypes.number,
};

IdleLoansList.defaultProps = {
  showMaxEntries: 5,
};
