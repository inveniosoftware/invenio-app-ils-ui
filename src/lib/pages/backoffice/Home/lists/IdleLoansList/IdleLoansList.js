import { dateFormatter, toShortDate } from '@api/date';
import { loanApi } from '@api/loans';
import { SeeAllButton } from '@components/backoffice/buttons/SeeAllButton';
import { Error } from '@components/Error';
import { Loader } from '@components/Loader';
import { ResultsTable } from '@components/ResultsTable/ResultsTable';
import { invenioConfig } from '@config';
import { BackOfficeRoutes } from '@routes/urls';
import { DateTime } from 'luxon';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Button } from 'semantic-ui-react';

export default class IdleLoansList extends Component {
  componentDidMount() {
    const { fetchIdlePendingLoans } = this.props;
    fetchIdlePendingLoans();
  }

  seeAllButton = () => {
    const path = BackOfficeRoutes.loansListWithQuery(
      loanApi
        .query()
        .withState(invenioConfig.CIRCULATION.loanRequestStates)
        .withUpdated({ to: toShortDate(DateTime.local().minus({ days: 10 })) })
        .qs()
    );
    return <SeeAllButton to={path} />;
  };

  viewDetails = ({ row }) => {
    return (
      <Button
        as={Link}
        to={BackOfficeRoutes.loanDetailsFor(row.metadata.pid)}
        compact
        icon="info"
        data-test={row.metadata.pid}
      />
    );
  };

  renderTable(data) {
    const { showMaxEntries } = this.props;
    const columns = [
      { title: '', field: '', formatter: this.viewDetails },
      { title: 'ID', field: 'metadata.pid' },
      { title: 'Patron ID', field: 'metadata.patron_pid' },
      { title: 'Document ID', field: 'metadata.document_pid' },
      {
        title: 'Request start date',
        field: 'metadata.request_start_date',
        formatter: dateFormatter,
      },
    ];
    return (
      <ResultsTable
        data={data.hits}
        columns={columns}
        totalHitsCounts={data.total}
        title="Idle loan requests"
        subtitle="Loan requests pending since more than 10 days."
        name="idle loan requests"
        seeAllComponent={this.seeAllButton()}
        showMaxRows={showMaxEntries}
      />
    );
  }

  render() {
    const { data, isLoading, error } = this.props;
    return (
      <Loader isLoading={isLoading}>
        <Error error={error}>{this.renderTable(data)}</Error>
      </Loader>
    );
  }
}

IdleLoansList.propTypes = {
  showMaxEntries: PropTypes.number,
  /* redux */
  fetchIdlePendingLoans: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired,
  error: PropTypes.object,
};

IdleLoansList.defaultProps = {
  showMaxEntries: 5,
  error: {},
};
