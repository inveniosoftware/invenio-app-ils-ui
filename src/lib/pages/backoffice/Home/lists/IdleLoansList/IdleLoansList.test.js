import { BackOfficeRoutes } from '@routes/urls';
import testData from '@testData/loans.json';
import { mount, shallow } from 'enzyme';
import React from 'react';
import IdleLoansList from './IdleLoansList';

jest.mock('react-router-dom');
jest.mock('@config');
BackOfficeRoutes.loanDetailsFor = jest.fn((pid) => `url/${pid}`);
let mockViewDetails = jest.fn();

const data = {
  hits: [
    {
      id: 1,
      pid: 'loan1',
      metadata: testData[0],
    },
    {
      id: 2,
      pid: 'loan2',
      metadata: testData[1],
    },
  ],
  total: 2,
};

let component;
afterEach(() => {
  if (component) {
    mockViewDetails.mockClear();
    component.unmount();
  }
});

describe('IdleLoansList tests', () => {
  it('should load the details component', () => {
    const component = shallow(
      <IdleLoansList
        data={{ hits: [], total: 0 }}
        isLoading={false}
        fetchIdlePendingLoans={() => {}}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should fetch loans on mount', () => {
    const mockedFetchLoans = jest.fn();
    component = mount(
      <IdleLoansList
        data={{ hits: [], total: 0 }}
        isLoading={false}
        fetchIdlePendingLoans={mockedFetchLoans}
      />
    );
    expect(mockedFetchLoans).toHaveBeenCalled();
  });

  it('should render show a message with no loans', () => {
    component = mount(
      <IdleLoansList
        data={{ hits: [], total: 0 }}
        isLoading={false}
        fetchIdlePendingLoans={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
    const message = component
      .find('Message')
      .filterWhere((element) => element.prop('data-test') === 'no-results');
    expect(message).toHaveLength(1);
  });
});
