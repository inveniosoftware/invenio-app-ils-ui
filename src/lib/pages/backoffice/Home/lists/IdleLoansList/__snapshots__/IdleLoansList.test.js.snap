// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdleLoansList tests should load the details component 1`] = `
<Overridable(Loader)
  isLoading={false}
>
  <Error
    FallbackComponent={[Function]}
    boundary={false}
    error={Object {}}
    onUIError={null}
  >
    <ResultsTable
      columns={
        Array [
          Object {
            "field": "",
            "formatter": [Function],
            "title": "",
          },
          Object {
            "field": "metadata.pid",
            "title": "ID",
          },
          Object {
            "field": "metadata.patron_pid",
            "title": "Patron ID",
          },
          Object {
            "field": "metadata.document_pid",
            "title": "Document ID",
          },
          Object {
            "field": "metadata.request_start_date",
            "formatter": [Function],
            "title": "Request start date",
          },
        ]
      }
      currentPage={1}
      data={Array []}
      fixed={false}
      headerActionComponent={null}
      name="idle loan requests"
      paginationComponent={null}
      renderEmptyResultsElement={null}
      seeAllComponent={
        <SeeAllButton
          disabled={false}
          fluid={false}
          to="/backoffice/loans?q=(state:(PENDING) AND _updated:%7B*%20TO%202019-10-18%7D)"
        />
      }
      showAllResults={false}
      showMaxRows={5}
      singleLine={true}
      subtitle="Loan requests pending since more than 10 days."
      title="Idle loan requests"
      totalHitsCount={0}
      totalHitsCounts={0}
    />
  </Error>
</Overridable(Loader)>
`;

exports[`IdleLoansList tests should render show a message with no loans 1`] = `
<IdleLoansList
  data={
    Object {
      "hits": Array [],
      "total": 0,
    }
  }
  error={Object {}}
  fetchIdlePendingLoans={[Function]}
  isLoading={false}
  showMaxEntries={5}
>
  <Overridable(Loader)
    isLoading={false}
  >
    <Loader
      isLoading={false}
    >
      <Overridable
        id="Loader.layout"
        isLoading={false}
      >
        <Error
          FallbackComponent={[Function]}
          boundary={false}
          error={Object {}}
          onUIError={null}
        >
          <ResultsTable
            columns={
              Array [
                Object {
                  "field": "",
                  "formatter": [Function],
                  "title": "",
                },
                Object {
                  "field": "metadata.pid",
                  "title": "ID",
                },
                Object {
                  "field": "metadata.patron_pid",
                  "title": "Patron ID",
                },
                Object {
                  "field": "metadata.document_pid",
                  "title": "Document ID",
                },
                Object {
                  "field": "metadata.request_start_date",
                  "formatter": [Function],
                  "title": "Request start date",
                },
              ]
            }
            currentPage={1}
            data={Array []}
            fixed={false}
            headerActionComponent={null}
            name="idle loan requests"
            paginationComponent={null}
            renderEmptyResultsElement={null}
            seeAllComponent={
              <SeeAllButton
                disabled={false}
                fluid={false}
                to="/backoffice/loans?q=(state:(PENDING) AND _updated:%7B*%20TO%202019-10-18%7D)"
              />
            }
            showAllResults={false}
            showMaxRows={5}
            singleLine={true}
            subtitle="Loan requests pending since more than 10 days."
            title="Idle loan requests"
            totalHitsCount={0}
            totalHitsCounts={0}
          >
            <Grid>
              <div
                className="ui grid"
              >
                <GridRow>
                  <div
                    className="row"
                  >
                    <GridColumn
                      verticalAlign="middle"
                      width={16}
                    >
                      <div
                        className="middle aligned sixteen wide column"
                      >
                        <Header
                          as="h3"
                          content="Idle loan requests"
                          subheader="Loan requests pending since more than 10 days."
                        >
                          <h3
                            className="ui header"
                          >
                            Idle loan requests
                            <HeaderSubheader
                              content="Loan requests pending since more than 10 days."
                            >
                              <div
                                className="sub header"
                              >
                                Loan requests pending since more than 10 days.
                              </div>
                            </HeaderSubheader>
                          </h3>
                        </Header>
                      </div>
                    </GridColumn>
                    <GridColumn
                      textAlign="right"
                      width={8}
                    >
                      <div
                        className="right aligned eight wide column"
                      />
                    </GridColumn>
                  </div>
                </GridRow>
              </div>
            </Grid>
            <Message
              data-test="no-results"
              icon={true}
              info={true}
            >
              <div
                className="ui icon info message"
                data-test="no-results"
              >
                <MessageContent>
                  <div
                    className="content"
                  >
                    <MessageHeader>
                      <div
                        className="header"
                      >
                        No results found
                      </div>
                    </MessageHeader>
                    There are no 
                    idle loan requests
                    .
                  </div>
                </MessageContent>
              </div>
            </Message>
          </ResultsTable>
        </Error>
      </Overridable>
    </Loader>
  </Overridable(Loader)>
</IdleLoansList>
`;
